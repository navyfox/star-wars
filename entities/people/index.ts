export * from './people';

// export default slice.peopleSliceReducer;
