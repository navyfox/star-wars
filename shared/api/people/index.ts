export * from './people';
export * from './types';
