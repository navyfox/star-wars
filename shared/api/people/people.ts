import axios from 'axios';
import { Person, PeopleApiResponse } from './types';

export const getPeopleByPage = async (page: number | string) =>
  axios
    .get(`${process.env.NEXT_PUBLIC_STAR_WARS_API}/people/?page=${page}`, {
      headers: {
        'Content-Type': 'application/json',
      },
    })
    .then((resp) => resp.data as PeopleApiResponse);

export const getPersonByID = async (id: number | string) =>
  axios.get(`${process.env.NEXT_PUBLIC_STAR_WARS_API}/people/${id}`, {
    headers: {
      'Content-Type': 'application/json',
    },
  });
