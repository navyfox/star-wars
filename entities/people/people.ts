import { createSlice, createAction, createAsyncThunk } from '@reduxjs/toolkit';

import { HYDRATE } from 'next-redux-wrapper';
import { Person } from '@/shared/api/people';

// mock
const savePersonApi = (person: Person | null) =>
  new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('SAVE_PERSON'), person;
      resolve({ isSaving: true });
    }, 1000);
  });

export interface PeopleState {
  person: Person | null;
  isSaving: boolean;
  isPending: boolean;
  error: string | null;
}

const initialState: PeopleState = {
  person: null,
  isSaving: true,
  isPending: false,
  error: null,
};

const hydrate = createAction<{ people: PeopleState }>(HYDRATE);

export const savePerson = createAsyncThunk(
  'people/savePerson',
  async (person: Person | null, { rejectWithValue }) => {
    try {
      const response = await savePersonApi(person);
      return response;
    } catch (error: any) {
      return rejectWithValue(error.response.data);
    }
  }
);

export const peopleSlice = createSlice({
  name: 'people',
  initialState,
  reducers: {
    setPerson(state, action) {
      state.person = { ...state.person, ...action.payload };
    },
    setIsSaving(state, action) {
      state.isSaving = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(hydrate, (state, action) => {
      return {
        ...state,
        ...action.payload.people,
      };
    });
    builder.addCase(savePerson.pending, (state) => {
      state.isPending = true;
    });
    builder.addCase(savePerson.fulfilled, (state, action) => {
      state.isSaving = (action.payload as { isSaving: boolean }).isSaving;
    });
    builder.addCase(savePerson.rejected, (state, action) => {
      state.isSaving = false;
      state.error = action.payload as string;
    });
  },
});

export const { setPerson, setIsSaving } = peopleSlice.actions;

export default peopleSlice.reducer;
