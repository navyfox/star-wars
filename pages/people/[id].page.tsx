import React, { ChangeEvent } from 'react';
import Head from 'next/head';
import { GetServerSideProps } from 'next';
import Container from '@mui/material/Container';
import { useSelector, useDispatch } from 'react-redux';
import { setPerson } from '@/entities/people';
import { wrapperStore } from '@/settings/store';
import Box from '@mui/material/Box';
import PersonEditCard from '@/widgets/person-edit-card';
import { getPersonByID } from '@/shared/api/people';
import { Person } from '@/shared/api/people';
import { savePerson, setIsSaving } from '@/entities/people';
import Button from '@mui/material/Button';
import WrapperCard from '@/shared/ui/wrapper-card';
import Stack from '@mui/material/Stack';
import type { NextPage } from 'next/types';

export const PersonPage: NextPage = () => {
  const dispatch = useDispatch<AppThunk>();
  const person = useSelector((state: AppState) => state.people.person);

  const onSave = () => {
    dispatch(savePerson(person));
  };

  const onChangePerson = (event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    dispatch(
      setPerson({
        [event.target.id]: event.target.value,
      })
    );
    dispatch(setIsSaving(false));
  };
  return (
    <>
      <Head>
        <title>People from Star Wars</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
      </Head>
      <Container maxWidth="sm">
        <Box marginTop={5} data-testid="person-card">
          <WrapperCard>
            <PersonEditCard person={person} onChangePerson={onChangePerson} />
            <Stack marginTop={5} marginBottom={1} alignItems="center">
              <Button variant="outlined" onClick={onSave}>
                SAVE
              </Button>
            </Stack>
          </WrapperCard>
        </Box>
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = wrapperStore.getServerSideProps(
  (store) =>
    async ({ query: { id = '1' } }) => {
      const pageIndex = Array.isArray(id) ? parseInt(id[0]) : parseInt(id) || 1;
      const person = await getPersonByID(pageIndex).then((resp) => resp.data as Person);
      await store.dispatch(setPerson(person));
      return { props: {} };
    }
);

export default PersonPage;
