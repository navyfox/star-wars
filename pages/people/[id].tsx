import React, { ChangeEvent } from 'react';
import Head from 'next/head';
import { GetServerSideProps } from 'next';
import Container from '@mui/material/Container';
import { useSelector, useDispatch } from 'react-redux';
import { setPerson } from '@/entities/people';
import { wrapperStore } from '@/app/store';
import Box from '@mui/material/Box';
import PersonEditCard from '@/widgets/person-edit-card';
import { getPersonByID } from '@/shared/api/people';
import { Person } from '@/shared/api/people';

export const PersonPage = () => {
  const dispatch = useDispatch();
  const person = useSelector((state: AppState) => state.people.person);

  const onChangePerson = (event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
    dispatch(
      setPerson({
        [event.target.id]: event.target.value,
      })
    );
  };
  return (
    <>
      <Head>
        <title>People from Star Wars</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
      </Head>
      <Container maxWidth="sm">
        <Box marginTop={5}>
          <PersonEditCard person={person} onChangePerson={onChangePerson} />
        </Box>
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = wrapperStore.getServerSideProps(
  (store) =>
    async ({ query: { id = '1' } }) => {
      const pageIndex = Array.isArray(id) ? parseInt(id[0]) : parseInt(id) || 1;
      const person = await getPersonByID(pageIndex).then((resp) => resp.data as Person);
      await store.dispatch(setPerson(person));
      return { props: {} };
    }
);

export default PersonPage;
