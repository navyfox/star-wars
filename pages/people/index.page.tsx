import React from 'react';
import type { NextPage } from 'next/types';
import Head from 'next/head';
import { GetServerSideProps } from 'next';
import Container from '@mui/material/Container';
import { getPeopleByPage } from '@/shared/api/people';
import { Person } from '@/shared/api/people';
import PeoplePagination from '@/widgets/people-pagination';

interface PeoplePageProps {
  count: number;
  people: Person[];
  page: number;
}

export const PeoplePage: NextPage<PeoplePageProps> = ({ count, people, page }: PeoplePageProps) => {
  return (
    <>
      <Head>
        <title>People from Star Wars</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
      </Head>
      <Container maxWidth="sm">
        <PeoplePagination count={count} people={people} page={page} />
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = async ({ query: { page = '1' } }) => {
  const pageIndex = Array.isArray(page) ? parseInt(page[0]) : parseInt(page) || 1;
  const data = await getPeopleByPage(pageIndex);

  return { props: { people: data.results, count: data.count, page: pageIndex } };
};

export default PeoplePage;
