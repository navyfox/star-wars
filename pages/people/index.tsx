import React from 'react';
import Head from 'next/head';
import { GetServerSideProps } from 'next';
import Container from '@mui/material/Container';
import { getPeopleByPage } from '@/shared/api/people';
import { Person } from '@/shared/api/people';
import PeoplePagination from '@/widgets/people-pagination';

export const PeoplePage = ({
  count,
  people,
  page,
}: {
  count: number;
  people: Person[];
  page: number;
}) => {
  return (
    <>
      <Head>
        <title>People from Star Wars</title>
        <meta name="viewport" content="width=device-width, initial-scale=1" />
      </Head>
      <Container maxWidth="sm">
        <PeoplePagination count={count} people={people} page={page} />
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = async ({ query: { page = '1' } }) => {
  const pageIndex = Array.isArray(page) ? parseInt(page[0]) : parseInt(page) || 1;
  const data = await getPeopleByPage(pageIndex);

  return { props: { people: data.results, count: data.count, page: pageIndex } };
};

export default PeoplePage;
