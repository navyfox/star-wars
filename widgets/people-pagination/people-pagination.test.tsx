import React from 'react';
import { render } from '@testing-library/react';
import { Person } from '@/shared/api/people';
import renderer from 'react-test-renderer';
import PeoplePagination from './index';

const mockPeople: Person[] = [
  {
    name: 'Luke Skywalker',
    height: '172',
    mass: '77',
    gender: 'male',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    url: 'https://swapi.dev/api/people/1/',
  },
  {
    name: 'Leia Organa',
    height: '150',
    mass: '49',
    gender: 'female',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    url: 'https://swapi.dev/api/people/5/',
  },
];

describe('PeoplePagination', () => {
  it('renders correctly', () => {
    const component = renderer
      .create(<PeoplePagination count={mockPeople.length} people={mockPeople} page={1} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders the correct number of people cards', async () => {
    const { queryAllByTestId } = await render(
      <PeoplePagination count={mockPeople.length} people={mockPeople} page={1} />
    );
    expect(await queryAllByTestId('person-card', { exact: false })).toHaveLength(mockPeople.length);
  });

  it('renders pagination with correct page count', () => {
    const pageSize = 10;
    const { queryAllByTestId } = render(
      <PeoplePagination count={30} people={mockPeople} page={1} />
    );
    const pagination = queryAllByTestId('pages');
    expect(pagination).toHaveLength(Math.ceil(30 / pageSize) + 2);
  });
});
