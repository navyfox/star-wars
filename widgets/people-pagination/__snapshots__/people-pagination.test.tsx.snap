// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeoplePagination renders correctly 1`] = `
[
  <div
    className="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-1ho1yxb-MuiGrid2-root"
    id="wrap"
  >
    <div
      className="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-sm-6 css-luttq2-MuiGrid2-root"
    >
      <a
        data-testid="person-card"
        href="https://swapi.dev/api/people/1/"
        id="person-1"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "textDecoration": "none",
          }
        }
        target="_blank"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1gictic-MuiPaper-root"
        >
          <div
            className="MuiBox-root css-0"
          >
            name: 
            Luke Skywalker
          </div>
          <div
            className="MuiBox-root css-0"
          >
            gender: 
            male
          </div>
          <div
            className="MuiBox-root css-0"
          >
            height: 
            172
          </div>
          <div
            className="MuiBox-root css-0"
          >
            mass: 
            77
          </div>
        </div>
      </a>
    </div>
    <div
      className="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-sm-6 css-luttq2-MuiGrid2-root"
    >
      <a
        data-testid="person-card"
        href="https://swapi.dev/api/people/5/"
        id="person-5"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "textDecoration": "none",
          }
        }
        target="_blank"
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1gictic-MuiPaper-root"
        >
          <div
            className="MuiBox-root css-0"
          >
            name: 
            Leia Organa
          </div>
          <div
            className="MuiBox-root css-0"
          >
            gender: 
            female
          </div>
          <div
            className="MuiBox-root css-0"
          >
            height: 
            150
          </div>
          <div
            className="MuiBox-root css-0"
          >
            mass: 
            49
          </div>
        </div>
      </a>
    </div>
  </div>,
  <div
    className="MuiStack-root css-yzlqr0-MuiStack-root"
  >
    <nav
      aria-label="pagination navigation"
      className="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
    >
      <ul
        className="MuiPagination-ul css-wjh20t-MuiPagination-ul"
      >
        <li>
          <a
            aria-disabled={true}
            aria-label="Go to previous page"
            className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            data-testid="pages"
            href="/people?page=0"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
              data-testid="NavigateBeforeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </a>
        </li>
        <li>
          <a
            aria-current="true"
            aria-label="page 1"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-selected MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            data-testid="pages"
            href="/people"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
          >
            1
          </a>
        </li>
        <li>
          <a
            aria-disabled={true}
            aria-label="Go to next page"
            className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            data-testid="pages"
            href="/people?page=2"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
              data-testid="NavigateNextIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </a>
        </li>
      </ul>
    </nav>
  </div>,
]
`;
